### PR TITLE
Derive kairos chunk sizes

### PR DESCRIPTION
**Pull request by zachm**
This is part of @venkey-ariv's proposal to chunk up Kairos queries. The basic idea is to split a 3h query, for example, into six 30m queries. This can radically speed up Kairos backend response times by sharding out the work.

The actual derivation is a little complicated. We add a default chunking size, and a maximum number of chunks we're willing to proxy out, into the config. The implementation respects this and tries to do the right thing.

I wrote tests - full coverage - because there are many edge cases and I'd prefer to not introduce more bugs.

### CODE REVIEW

**Comment by venkey-ariv** [comment]
I tested the chunking logic in an ad-hoc fashion. This is result when I issued -1h to now. 
1. 2016-03-09 02:44:36.855334 2016-03-09 03:14:36.855334
2. 2016-03-09 02:14:36.855334 2016-03-09 02:44:36.855334
3. 2016-03-09 02:14:36.855324 2016-03-09 02:14:36.855334

The end time of a window and the start time of a window seem to overlap. endtime of 2 is the start time of 1. Wouldn't this cause extra data points to be queried out? or am I sleep deprived? I think 2 should be:

2016-03-09 02:14:36.855334 2016-03-09 02:44:35.xxxxxx

Also 3's start and end times are the same - i understand why this is happening. But can we drop the last value in the list before returning it to the caller if time difference is <1s?

File: tests/test_utils.py [hunk 1]
```diff
@@ -125,6 +126,53 @@ def test_get_needed_absolute_time_range(m_dt):
...
+    results = get_chunked_time_ranges({'chunking': {}}, kairos_timing)
+    assert len(results) == 6
+    for i in xrange(len(results)):
+        offset = datetime.timedelta(minutes=(i * 120))
+        assert results[i][1] == now - offset
+        assert results[i][0] == now - offset - datetime.timedelta(minutes=120)
+
+
+@freeze_time("2016-01-01 20:00:00", tz_offset=-8)
+def test_get_chunked_time_ranges_last_5h15m():
```
Comment (tiras-j) [inline tests/test_utils.py]: nit but shouldn't this be "...ranges_last_2h15m" since the time is relative 135minutes?

**Comment by zachm** [comment]
@venkey-ariv : excellent catches!

The extra tiny chunk was caused by drift between calls to `datetime.datetime.now()`. I've patched that and added a test that does _not_ freeze time, and instead uses a fuzzy match. If the box takes over two seconds to run the test it might fail... but best I can do given the circumstance.

I also cut down precision to second throughout - many other parts of tscached only work at second level due to timestamp handling - and added a 1s offset on start times. This causes adjacent to not share endpoints but instead have adjacent ones. 

``` python
from tscached.utils import get_chunked_time_ranges

query = {'start_relative': {'unit': 'hours', 'value': '1'}}
results = get_chunked_time_ranges({'chunking': {}}, query)

print len(results)
for result in results:
    print result[0], result[1]
```

Which gives me:

```
2
2016-03-09 10:23:40 2016-03-09 10:53:39
2016-03-09 09:53:40 2016-03-09 10:23:39
```

Oh and @tiras-j I renamed that test. Thanks :)

File: tscached/utils.py [hunk 2]
```diff
@@ -105,6 +111,52 @@ def get_needed_absolute_time_range(time_range):
...
+        # end_time is mutated below here
+        for i in xrange(num_chunks):
+            start = end_time - length_td
+
+            # Sanity check: make sure we limit the earliest chunk if it's partial
+            if start < start_time:
+                start = start_time
+
+            # Add an offset so the chunks don't overlap.
+            start += datetime.timedelta(seconds=1)
```
Comment (tiras-j) [inline tscached/utils.py:153]: Another nit, but wouldn't this introduce the possibility of missing a data point if it happens to fall into the 1 second gap between chunks? Why not add 1ms, or 1us, whatever the kairos fidelity is? That way we guarantee both not missing data points between chunks and not getting duplicate data points returned from kairos on the off chance there is a data point at the exact timestamp of the chunk split.

Comment (tiras-j) [inline tscached/utils.py:153]: Oh or is 1 sec the highest fidelity that kairos handles? in which case completely ignore my previous comment

**Comment by tiras-j** [comment]
+1 lgtm, my comments are just for consideration.

**Comment by venkey-ariv** [comment]
Ship it! :)
